acp/acp_local.go

```diff
@@ -0,0 +1,323 @@
+		return newErrResourceDoesNotExistOnTargetPolicy(err, resourceName, policyID)
+	}
+
+	// Now that we have validated that policyID exists and it contains a corresponding
+	// resource with the matching name, validate that all required permissions
+	// for DPI actually exist on the target resource.
+	for _, requiredPermission := range dpiRequiredPermissions {
+		permissionResponse := resourceResponse.GetPermissionByName(requiredPermission)
+		if permissionResponse == nil {
+			return newErrResourceIsMissingRequiredPermission(
+				err,
+				resourceName,
+				requiredPermission,
+				policyID,
+			)
+		}
+
+		// Now we need to ensure that the "owner" relation has access to all the required
+		// permissions for DPI. This is important because even if the policy has the required
+		// permissions under the resource, it's possible that those permissions are not granted
+		// to the "owner" relation, this will help users not shoot themseleves in the foot.
+		// TODO-ACP: Better validation, once sourcehub implements meta-policies.
+		// Issue: https://github.com/sourcenetwork/defradb/issues/2359
+		if err := validateDPIExpressionOfRequiredPermission(
+			permissionResponse.Expression,
+			requiredPermission,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (l *ACPLocal) RegisterDocObject(
+	ctx context.Context,
+	actorID string,
+	policyID string,
+	resourceName string,
+	docID string,
+) error {
+	registerDoc := types.MsgRegisterObject{
+		Creator:      actorID,
+		PolicyId:     policyID,
+		Object:       types.NewObject(resourceName, docID),
+		CreationTime: protoTypes.TimestampNow(),
+	}
+
+	registerDocResponse, err := l.localModule.GetMsgService().RegisterObject(
+		l.localModule.GetCtx(),
+		&registerDoc,
+	)
+
+	if err != nil {
+		return NewErrFailedToRegisterDocWithACP(err, "Local", policyID, actorID, resourceName, docID)
+	}
+
+	switch registerDocResponse.Result {
+	case types.RegistrationResult_NoOp:
+		return ErrObjectDidNotRegister
+
+	case types.RegistrationResult_Registered:
+		log.InfoContext(
+			ctx,
+			"Document registered with local acp module",
+			corelog.Any("PolicyID", policyID),
+			corelog.Any("Creator", actorID),
+			corelog.Any("Resource", resourceName),
+			corelog.Any("DocID", docID),
+		)
+		return nil
+
+	case types.RegistrationResult_Unarchived:
+		log.InfoContext(
+			ctx,
+			"Document re-registered (unarchived object) with local acp module",
+			corelog.Any("PolicyID", policyID),
+			corelog.Any("Creator", actorID),
+			corelog.Any("Resource", resourceName),
+			corelog.Any("DocID", docID),
+		)
+		return nil
+	}
+
+	return ErrObjectDidNotRegister
+}
+
+func (l *ACPLocal) IsDocRegistered(
+	ctx context.Context,
+	policyID string,
+	resourceName string,
+	docID string,
+) (bool, error) {
+	queryObjectOwner := types.QueryObjectOwnerRequest{
+		PolicyId: policyID,
+		Object:   types.NewObject(resourceName, docID),
+	}
+
+	queryObjectOwnerResponse, err := l.localModule.GetQueryService().ObjectOwner(
```

Comment:
I'm not debating the concept. I'm simply pointing out that the name of the methods create confusion.

Reply:
Since this is just a thought and discussion regarding a sourcehub function I am going to resolve this conversation. Feel free to re-raise if there is a suggestion or any follow-ups.